dali/kernels/signal/dct/dct_cpu.cc

```diff
@@ -0,0 +1,214 @@
+  }
+}
+
+template <typename T>
+void FillCosineTableTypeII(T *table, int64_t input_length, int64_t ndct, bool normalize) {
+  T phase_mul = M_PI / input_length;
+  T factor_k_0 = 1, factor_k_i = 1;
+  if (normalize) {
+    factor_k_i = std::sqrt(2 / T(input_length));
+    factor_k_0 = factor_k_i / std::sqrt(T(2));
+  }
+  int64_t idx = 0;
+  for (int64_t k = 0; k < ndct; k++) {
+    T norm_factor = (k == 0) ? factor_k_0 : factor_k_i;
+    for (int64_t n = 0; n < input_length; n++) {
+      table[idx++] = norm_factor * std::cos(phase_mul * (n + T(0.5)) * k);
+    }
+  }
+}
+
+
+template <typename T>
+void FillCosineTableTypeIII(T *table, int64_t input_length, int64_t ndct, bool normalize) {
+  T phase_mul = M_PI / input_length;
+  T factor_n_0 = 0.5, factor_n_i = 1;
+  if (normalize) {
+    factor_n_i = std::sqrt(T(2) / input_length);
+    factor_n_0 = factor_n_i / std::sqrt(T(2));
+  }
+  int64_t idx = 0;
+  for (int64_t k = 0; k < ndct; k++) {
+    table[idx++] = factor_n_0;  // n = 0
+    for (int64_t n = 1; n < input_length; n++) {
+      table[idx++] = factor_n_i * std::cos(phase_mul * n * (k + T(0.5)));
+    }
+  }
+}
+
+
+template <typename T>
+void FillCosineTableTypeIV(T *table, int64_t input_length, int64_t ndct, bool normalize) {
+  T phase_mul = M_PI / input_length;
+  T factor = normalize ? std::sqrt(T(2)/input_length) : T(1);
+  int64_t idx = 0;
+  for (int64_t k = 0; k < ndct; k++) {
+    for (int64_t n = 0; n < input_length; n++) {
+      table[idx++] = factor * std::cos(phase_mul * (n + T(0.5)) * (k + T(0.5)));
+    }
+  }
+}
+
+
+template <typename T>
+void FillCosineTable(T *table, int64_t input_length, int64_t ndct, int dct_type, bool normalize) {
+  switch (dct_type) {
+    case 1:
+      FillCosineTableTypeI(table, input_length, ndct, normalize);
+      break;
+    case 2:
+      FillCosineTableTypeII(table, input_length, ndct, normalize);
+      break;
+    case 3:
+      FillCosineTableTypeIII(table, input_length, ndct, normalize);
+      break;
+    case 4:
+      FillCosineTableTypeIV(table, input_length, ndct, normalize);
+      break;
+    default:
+      assert(false);
+  }
+}
+
+}  // namespace
+
+template <typename OutputType, typename InputType, int Dims>
+Dct1DCpu<OutputType, InputType, Dims>::~Dct1DCpu() = default;
+
+template <typename OutputType, typename InputType, int Dims>
+KernelRequirements
+Dct1DCpu<OutputType, InputType, Dims>::Setup(KernelContext &context,
+                                             const InTensorCPU<InputType, Dims> &in,
+                                             const DctArgs &original_args) {
+  const auto &in_shape = in.shape;
+  DALI_ENFORCE(in_shape.size() == Dims);
+
+  auto args = original_args;
+  args.axis = args.axis >= 0 ? args.axis : Dims - 1;
+  DALI_ENFORCE(args.axis >= 0 && args.axis < Dims,
+               make_string("Axis is out of bounds: ", args.axis));
+  int64_t n = in.shape[args.axis];
+
+  if (args.dct_type == 1) {
```

Comment:
We have that check repeated in `FillCosineTableTypeI`. Do we need to have that in two places?

Reply:
It's ok - this is a public API, so we throw - in FillCosineTable we assert (it should not be reachable with such parameter, regardless of what external caller does)..